### ISSUE TITLE
Allow unused protocol versions to be discarded at link-time

### ISSUE THREAD

**Issue by ctz** [ctz/rustls#224]
If you were only interested in using TLS1.3, it would be ideal if all the TLS1.2 code were discarded at link-time. For code size, complexity and very convincing attack surface reduction reasons.

**Comment by Ralith** [comment]
This would be awesome for QUIC!

**Comment by briansmith** [comment]
> If you were only interested in using TLS1.3, it would be ideal if all the TLS1.2 code were discarded at link-time. For code size, complexity and very convincing attack surface reduction reasons.

I don't think we need to worry too much about being able to discard TLS-1.3-specific code at link time. I suggest we optimize this work for the case where TLS 1.3 is enabled, and focus on making sure as little TLS-1.2-specific code gets linked in as practical. 

The "tls12" feature provides a way for people to statically avoid the vast majority of the TLS-1.2-specific code. It would be nice to have `SupportedProtocolVersion` become a trait, where the code protected by `#[cfg(feature = "tls12")]` would be moved to the TLS 1.2 implementations of the methods of that trait. But, maybe the "tls12" feature is good enough.

**Comment by ctz** [comment]
> maybe the "tls12" feature is good enough.

Really, cargo features as a means for detailed conditional compilation doesn't work well. This is quite a mess:

```
$ ack 'feature = "tls12"' | wc -l
110
```

And it grows ever more complex.

**Comment by djc** [comment]
I wonder how hard it would be to turn `SupportedCipherSuite` into a trait instead of an enum, I think that would help quite a bit with this.
